src/lf_apply.rs

```diff
@@ -44,78 +46,60 @@ unsafe fn backup_lpf<BD: BitDepth>(
-                    &mut *dst.offset(BD::pxstride(dst_stride as usize * 2) as isize)
-                        as *mut BD::Pixel,
-                    dst_w as usize,
-                ),
-                slice::from_raw_parts(
-                    &mut *dst
-                        .offset(BD::pxstride(dst_stride as usize * (top + 2) as usize) as isize)
-                        as *mut BD::Pixel,
-                    dst_w as usize,
-                ),
-                dst_w as usize,
-            );
-            BD::pixel_copy(
-                slice::from_raw_parts_mut(
-                    &mut *dst.offset(BD::pxstride(dst_stride as usize * 3) as isize)
-                        as *mut BD::Pixel,
-                    dst_w as usize,
-                ),
-                slice::from_raw_parts(
-                    &mut *dst
-                        .offset(BD::pxstride(dst_stride as usize * (top + 3) as usize) as isize)
-                        as *mut BD::Pixel,
-                    dst_w as usize,
-                ),
-                dst_w as usize,
-            );
+            if dst_stride < 0 {
+                let mut dst_offset_delta1 = 3 * BD::pxstride(-dst_stride) as usize;
+                let dst_tmp = dst.split_at_mut(dst_offset - dst_offset_delta1);
+                let mut dst_offset_delta0 =
+                    dst_offset - top as usize * BD::pxstride(-dst_stride) as usize;
+                dst_tmp.1[dst_offset_delta1..][..dst_w as usize]
+                    .copy_from_slice(&dst_tmp.0[dst_offset_delta0..][..dst_w as usize]);
+                dst_offset_delta0 -= BD::pxstride(-dst_stride) as usize;
+                dst_offset_delta1 -= BD::pxstride(-dst_stride) as usize;
+                dst_tmp.1[dst_offset_delta1..][..dst_w as usize]
+                    .copy_from_slice(&dst_tmp.0[dst_offset_delta0..][..dst_w as usize]);
+                dst_offset_delta0 -= BD::pxstride(-dst_stride) as usize;
+                dst_offset_delta1 -= BD::pxstride(-dst_stride) as usize;
+                dst_tmp.1[dst_offset_delta1..][..dst_w as usize]
+                    .copy_from_slice(&dst_tmp.0[dst_offset_delta0..][..dst_w as usize]);
+                dst_offset_delta0 -= BD::pxstride(-dst_stride) as usize;
+                dst_offset_delta1 -= BD::pxstride(-dst_stride) as usize;
+                dst_tmp.1[dst_offset_delta1..][..dst_w as usize]
+                    .copy_from_slice(&dst_tmp.0[dst_offset_delta0..][..dst_w as usize]);
+            } else {
+                let mut dst_offset_delta = 0;
+                let dst_tmp = dst.split_at_mut(
+                    (dst_offset as isize + top as isize * BD::pxstride(dst_stride)) as usize,
+                );
+                dst_tmp.0[dst_offset + dst_offset_delta..][..dst_w as usize]
+                    .copy_from_slice(&dst_tmp.1[dst_offset_delta..][..dst_w as usize]);
+                dst_offset_delta = (dst_offset_delta as isize + BD::pxstride(dst_stride)) as usize;
```

Comment:
Why do we have to do the addition through `isize` here?  `dst_stride >= 0` in this case, so it could be casted to `usize` and we can do everything through `usize`.

Reply:
I rewrote this section of code. Should be cleaner now and address this comment and others